### PR TITLE
Adding intersection meta

### PR DESCRIPTION
**Pull request by rorticus**
**Type:** feature

The following has been addressed in the PR:

* [ ] There is a related issue
* [x] All code matches the [style guide](https://github.com/dojo/meta/blob/master/STYLE.md)
* [x] Unit or Functional tests are included in the PR

**Description:**

Adding Intersection observer meta. This was originally split off from the meta PR and has been updated ...

* Based on latest meta
* Added `root` node support

### CODE REVIEW

File: src/meta/Intersection.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	private _rootObservers = new Map<string, IntersectionObserver>();
+
+	private _onIntersect(entries: IntersectionObserverEntry[], observer: IntersectionObserver) {
+		entries.forEach(entry => {
+			this._listeners.set(entry.target, entry);
+			this.invalidate();
+		});
+	}
+
+	get(key: string, { root = '' }: IntersectionOptions = {}): number {
```
Comment (agubler) [inline src/meta/Intersection.ts]: hello public visibility :)

Comment (rorticus) [inline src/meta/Intersection.ts]: ...

File: src/meta/Intersection.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+import WeakMap from '@dojo/shim/WeakMap';
+import MetaBase from '../meta/Base';
+
+import 'intersection-observer';
+
+export interface IntersectionOptions {
+	root?: string;
+}
+
+export default class Intersection extends MetaBase {
```
Comment (agubler) [inline src/meta/Intersection.ts]: generally we do a named export and default export for widget-core modules.

File: src/meta/Intersection.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+	private _listeners = new WeakMap<Element, IntersectionObserverEntry>();
+	private _rootObservers = new Map<string, IntersectionObserver>();
+
+	private _onIntersect(entries: IntersectionObserverEntry[], observer: IntersectionObserver) {
```
Comment (agubler) [inline src/meta/Intersection.ts]: What's the observer used for in this function?

File: src/meta/Intersection.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+		entries.forEach(entry => {
+			this._listeners.set(entry.target, entry);
+			this.invalidate();
+		});
+	}
+
+	public get(key: string, { root = '' }: IntersectionOptions = {}): number {
+		let rootObserver = this._rootObservers.get(root);
+		if (!rootObserver) {
+			const intersectionOptions: any = {
```
Comment (agubler) [inline src/meta/Intersection.ts]: is this really `any`?

File: src/meta/Intersection.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+		if (rootObserver) {
+			// is this node set up?
+			this.requireNode(key);
+			const node = this.nodes.get(key);
+			if (node) {
+				if (this._listeners.has(node)) {
+					return this._listeners.get(node).intersectionRatio;
+				}
+				else {
```
Comment (agubler) [inline src/meta/Intersection.ts]: don't need the else as already returned in the if

File: src/meta/Intersection.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+export interface IntersectionOptions {
+	root?: string;
+}
+
+export default class Intersection extends MetaBase {
+	private _listeners = new WeakMap<Element, IntersectionObserverEntry>();
+	private _rootObservers = new Map<string, IntersectionObserver>();
+
+	private _onIntersect(entries: IntersectionObserverEntry[], observer: IntersectionObserver) {
+		entries.forEach(entry => {
```
Comment (agubler) [inline src/meta/Intersection.ts]: the style guide says should use brackets for arrow functions `forEach((entry) => ...`

File: src/meta/Intersection.ts [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+			else {
+				rootObserver = new global.IntersectionObserver(this._onIntersect.bind(this), intersectionOptions);
+			}
+			if (rootObserver) {
+				this._rootObservers.set(root, rootObserver);
+			}
+		}
+
+		if (rootObserver) {
+			// is this node set up?
```
Comment (agubler) [inline src/meta/Intersection.ts]: do we need this comment?

**Comment by rorticus** [comment]
Closing this because @pottedmeat is making a faster strong better one.

**Pull request closed without merging**